libafl/src/inputs/bytessub.rs

```diff
@@ -0,0 +1,363 @@
+        assert_eq!(bsi.len(), 2);
+        bsi.bytes_mut()[0] = 3;
+        assert_eq!(bytes_input.bytes()[1], 3);
+
+        let mut bsi = bytes_input.sub_input(..);
+        assert_eq!(bsi.len(), len_orig);
+        bsi.bytes_mut()[0] = 1;
+        bsi.bytes_mut()[1] = 2;
+        assert_eq!(bytes_input.bytes()[0], 1);
+    }
+
+    #[test]
+    fn test_bytessubinput_resize() {
+        let (mut bytes_input, len_orig) = init_bytes_input();
+        let bytes_input_orig = bytes_input.clone();
+
+        let mut bsi = bytes_input.sub_input(2..);
+        assert_eq!(bsi.len(), len_orig - 2);
+        bsi.resize(len_orig, 0);
+        assert_eq!(bsi.bytes()[bsi.len() - 1], 0);
+        assert_eq!(bsi.len(), len_orig);
+        assert_eq!(bytes_input.len(), len_orig + 2);
+        assert_eq!(bytes_input.bytes()[bytes_input.len() - 1], 0);
+
+        let (mut bytes_input, len_orig) = init_bytes_input();
+
+        let mut bsi = bytes_input.sub_input(..2);
+        assert_eq!(bsi.len(), 2);
+        bsi.resize(3, 0);
+        assert_eq!(bsi.len(), 3);
+        println!("{bsi:?}");
+        assert_eq!(bsi.bytes()[bsi.len() - 1], 0);
+        assert_eq!(bytes_input.len(), len_orig + 1);
+
+        let mut bsi = bytes_input.sub_input(..3);
+        assert_eq!(bsi.len(), 3);
+        bsi.resize(2, 0);
+        assert_eq!(bsi.len(), 2);
+        println!("{bsi:?}");
+        assert_eq!(bytes_input, bytes_input_orig);
+
+        let mut bsi = bytes_input.sub_input(2..=2);
+        bsi.resize(2, 0);
+        println!("{bsi:?}");
+        bsi.resize(1, 0);
+        println!("{bsi:?}");
+        assert_eq!(bytes_input, bytes_input_orig);
+
+        let mut bsi = bytes_input.sub_input(..);
+        assert_eq!(bsi.len(), bytes_input_orig.len());
+        bsi.resize(1, 0);
+        assert_eq!(bsi.len(), 1);
+        bsi.resize(10, 0);
+        assert_eq!(bsi.len(), 10);
+        assert_eq!(bytes_input.len(), 10);
+        assert_eq!(bytes_input.bytes()[2], 0);
+
+        let mut bsi = bytes_input.sub_input(..);
+        bsi.resize(1, 0);
+        assert_eq!(bytes_input.len(), 1);
+    }
+
+    #[test]
+    fn test_bytessubinput_drain_extend() {
+        let (mut bytes_input, len_orig) = init_bytes_input();
+        let bytes_input_cloned = bytes_input.clone();
+
+        let mut bsi = bytes_input.sub_input(..2);
+        let drained: Vec<_> = bsi.drain(..).collect();
+        assert_eq!(bsi.len(), 0);
+        assert_eq!(bytes_input.len(), len_orig - 2);
+
+        let mut bsi = bytes_input.sub_input(..0);
+        assert_eq!(bsi.len(), 0);
+        let drained_len = drained.len();
+        bsi.extend(&drained[..]);
+        assert_eq!(bsi.len(), drained_len);
+        assert_eq!(bytes_input, bytes_input_cloned);
+    }
+
+    #[test]
+    fn test_bytessubinput_mutator() {
+        let (mut bytes_input, _len_orig) = init_bytes_input();
+        let bytes_input_cloned = bytes_input.clone();
+
+        let mut bsi = bytes_input.sub_input(..2);
```

Comment:
@riesentoaster does this help you?

Reply:
The issue is that my "main" `Input` still needs to implement `HasMutatorBytes`. I could just create a slice from all vectors in the struct concatenated, but that seems inefficient. Plus I'd then have to dynamically calculate the bounds by again querying the original `Input` for the lengths of the vector parts and calculate offsets.

For this to be much prettier, I'd need something similar, but which accepts a mapping/extracting closure instead of a simple index range, but can be implemented by any `input` without the requirement for `HasMutatorBytes`. Do you think that would be feasible?

Reply:
I don't fully understand what your goal is, but we can implement `HasMutatorBytes` for `Vec`, does that help?

Reply:
Not really. 

Some background: I have a custom `Input` type that has multiple binary entries (`Vec<u8>`), which are then mapped to command line parameters for my binary (coreutils in my case). I want to mutate them all individually according to the mutations defined in havoc_mutations.

Reply:
You want Multipart: https://github.com/AFLplusplus/LibAFL/blob/main/libafl/src/mutators/multi.rs

This is already default implemented, just change your input shape to `MultipartInput<BytesInput>`.

Reply:
My goal is to create something like [this](https://github.com/AFLplusplus/LibAFL/issues/2202#issuecomment-2119276999) (from #2202).

Reply:
Not 100% sure I understand, but what about if we implement `HasMutatorBytes` for `&mut Vec`?
Just added a PR, please take a look :)

Reply:
> My goal is to create something like [this](https://github.com/AFLplusplus/LibAFL/issues/2202#issuecomment-2119276999) (from #2202).

What you are describing is (I believe) exactly implemented by multipart...

Reply:
Otherwise, if it is not, you may use this PR to perform the mapping you describe, without unnecessary cloning.

Reply:
@addisoncrump oh, that might actually be a solution. The only issue is the following (from inputs/multi.rs):

> As such, [`MultipartInput`] requires that each subcomponent be the same subtype.

That is a problem, because not all command line arguments have parameters (such as `--help`). I also don't always want to pass the argument, so I'd need a construct like `MultipartInput<Option<CommandLineArg>>` with

```rust
enum CommandLineArg {
    WithParam(Vec<u8>),
    WithoutParam,
}
```

I'm not sure this is possible — is it?

Reply:
It would be possible with this PR, yeah :slightly_smiling_face: It would require you to apply the parameter and stdin mutations separately though, we don't have a good way to compose mutations presently.

Reply:
Not sure what you mean by that. I was just going to add a `CommandLineArg` with a fixed name for the stdin (yes, I should maybe rename it, or just use `Option<Option<Vec<u8>>>` since that's essentially what this is).

I'll wait for this PR to be merged and try again then. And write an issue or something if I can't figure it out.

Reply:
Why would you need to keep a `CommandLineArg` key in the input that doesn't have a value? Or do you also want to mutate the list of parameters themselves?

Reply:
Yes, exactly. I'm doing differential fuzzing and I want to restrict the parameter space to command line arguments that don't alter the environment (such as output file args). Within those, I want the fuzzer to explore based on different combinations of CLAs, including the presence/absence of them and different values for those that take parameters. So inputs passed to the executor (or how you'd pass them manually on the command line) could look something like this:

```bash
echo "This is a stdin byte array"           | binary_under_test --arg1_with_param "value of param for arg1" --arg2_without_param --arg3_without_param --arg4_with_param "some value"
echo "This is another byte array for stdin" | binary_under_test --arg1_with_param "mutated value of param for arg1"
echo "More stdin"                           | binary_under_test --arg2_without_param --arg3_without_param
```

Reply:
Sounds like a MutlipartInput will be the thing to use. Either way having the current PR in is good

Reply:
> Either way having the current PR in is good

I'll check again once this PR is merged. Thank you for your suggestions!